kafka-connect-s3/src/main/java/io/confluent/connect/s3/continuum/S3Continuum.java

```diff
@@ -100,19 +100,24 @@ public boolean isActive() {
     return this.producer != null;
   }
 
-  public void produce(String key, String filename, long offset, long recordCount) {
+  public void produce(String key,
+                      String filename,
+                      long startOffset,
+                      long endOffset,
+                      long recordCount) {
     if (isActive()) {
       boolean usingAvro = this.valueSchema != null;
       if (usingAvro) {
         GenericRecord value = new GenericData.Record(this.valueSchema);
         value.put("filename", filename);
-        value.put("offset", offset);
+        value.put("startOffset", startOffset);
```

Comment:
Do we need to update `s3NotificationSchema` above?

> I think this is technically a breaking change (unless schema registry handles this for us, which it might)

By default, schema registry [is in backwards compatibility mode](https://docs.confluent.io/platform/current/schema-registry/avro.html). Apparently that allows removing fields (like how we're removing offset) and adding "optional" fields. So we may need to somehow mark these new ones as optional/provide a default? Otherwise, schema registry might reject the schema. I guess we can test to see what happens if you send a broadcast using the same schema registry instance as a prior instance of the broadcast.

They also want you to upgrade the consumer before the producers so ensure the consumer can understand the new format prior to producers sending it.

Reply:
> Do we need to update s3NotificationSchema above?
Woops, we certainly do!

> Didn't the latest release include everything except the RabbitMQ part? So the schema registry will have the old version in it.

That's true. The most recent release had the sink publishing Kafka messages and the DFS consuming those messages and logging them to the console. I'll build an image out of this branch and do some manual testing in Minikube to make sure we can upgrade cleanly.

Reply:
> So we may need to somehow mark these new ones as optional/provide a default? Otherwise, schema registry might reject the schema.

I went with adding defaults, as it felt weird to mark the new stuff as optional if we're always expecting the fields to be populated. Both options are weird, but defaults felt less weird.

Reply:
Alright, here's what I did in Minikube:
1. Deployed the DFS with the existing KafkaConnect image (the one that publishes with the old schema).
2. Wrote some data (to make sure schema registry saw a message with the old schema).
3. Upgraded to the new KafkaConnect image with the new schema.
4. Wrote some data.

The DFS failed to consume the new message with a schema mismatch error, which is expected. We'll do a lockstep upgrade of the Avro parsing code when we uptake the new image. This suggests that the s3 sink didn't have any trouble publishing the message, which is what we were interested in. From the error that the DFS reported, I see the new startOffset and endOffset fields in the message, too:

```
{"timestamp":"2023-01-30T21:09:10.6297587Z","level":"warning","message":"Kafka Consumer - Error consuming, retrying in 00:00:01: Confluent.Kafka.ConsumeException: Local: Value deserialization error\n ---\u003E Avro.AvroException: Schema mismatch. Reader: {\u0022type\u0022:\u0022record\u0022,\u0022name\u0022:\u0022new_file\u0022,\u0022namespace\u0022:\u0022io.confluent.connect.s3.continuum\u0022,\u0022fields\u0022:[{\u0022name\u0022:\u0022filename\u0022,\u0022type\u0022:\u0022string\u0022},{\u0022name\u0022:\u0022offset\u0022,\u0022type\u0022:\u0022long\u0022},{\u0022name\u0022:\u0022recordCount\u0022,\u0022type\u0022:\u0022long\u0022}]}, writer: {\u0022type\u0022:\u0022record\u0022,\u0022name\u0022:\u0022new_file\u0022,\u0022namespace\u0022:\u0022io.confluent.connect.s3.continuum\u0022,\u0022fields\u0022:[{\u0022name\u0022:\u0022filename\u0022,\u0022type\u0022:\u0022string\u0022},{\u0022name\u0022:\u0022startOffset\u0022,\u0022default\u0022:0,\u0022type\u0022:\u0022long\u0022},{\u0022name\u0022:\u0022endOffset\u0022,\u0022default\u0022:0,\u0022type\u0022:\u0022long\u0022},{\u0022name\u0022:\u0022recordCount\u0022,\u0022type\u0022:\u0022long\u0022}]}\n   at Avro.Generic.DefaultReader.Read[T](T reuse, Decoder decoder)\n   at Avro.Specific.SpecificReader\u00601.Read(T reuse, Decoder dec)\n   at Confluent.SchemaRegistry.Serdes.SpecificDeserializerImpl\u00601.Deserialize(String topic, Byte[] array)\n   at Confluent.SchemaRegistry.Serdes.AvroDeserializer\u00601.DeserializeAsync(ReadOnlyMemory\u00601 data, Boolean isNull, SerializationContext context)\n   at Confluent.Kafka.SyncOverAsync.SyncOverAsyncDeserializer\u00601.Deserialize(ReadOnlySpan\u00601 data, Boolean isNull, SerializationContext context)\n   at Confluent.Kafka.Consumer\u00602.Consume(Int32 millisecondsTimeout)\n   --- End of inner exception stack trace ---\n   at Confluent.Kafka.Consumer\u00602.Consume(Int32 millisecondsTimeout)\n   at Confluent.Kafka.Consumer\u00602.Consume(TimeSpan timeout)\n   at NationalInstruments.DataFrameService.RowDataIngestion.Kafka.KafkaConsumer\u00601.ConsumeOnceAsync(IConsumer\u00602 consumer, KafkaConsumerCallbackAsync\u00601 callbackAsync, CancellationToken cancellationToken) in /app/DataFrameService/DataFrameService.RowDataIngestion/Kafka/KafkaConsumer.cs:line 121\n   at NationalInstruments.DataFrameService.RowDataIngestion.Kafka.KafkaConsumer\u00601.ConsumeUntilCanceledAsync(IConsumer\u00602 consumer, KafkaConsumerCallbackAsync\u00601 callbackAsync, Func\u00601 shouldConsume, CancellationToken cancellationToken) in /app/DataFrameService/DataFrameService.RowDataIngestion/Kafka/KafkaConsumer.cs:line 96"}
```

I think we should be good on the S3 sink side of things. There's a separate question of "should the new Avro parsing code in the DFS be able to read messages with the old schema too", and I'm not sure it's worth spending time being backwards compatible there if it doesn't come for free. It's unlikely that users will have messages with the old schema queued up when we do the upgrade. If that does happen, I think the workaround is to go into Kafka UI and clear the messages in the topic. Nobody's listening for our broadcasts yet, so it's safe to clear the messages.

Reply:
Nice. The schema registry document had mentioned updating consumers before producers. DFS is supposed to automatically throw away messages it fails to deserialize (Chris thought that was automatic without us needing to do anything). If it's not doing that, that's a bug. Otherwise, we shouldn't have to worry about clearing the topic or anything.

I guess if the new offset fields get marked as optional in DFS too, we may need to manually ignore messages that are missing the new fields.

Reply:
> DFS is supposed to automatically throw away messages it fails to deserialize (Chris thought that was automatic without us needing to do anything).

That is the behavior I think I'm seeing. Kafka UI reports both that no messages are queued, and that none of the consumers are behind. It's not totally clear from the DFS logs that we discarded a message, but we aren't stuck in a loop trying to consume stuff that we can't:

```
{"timestamp":"2023-01-30T21:04:10.6207758Z","level":"warning","message":"Kafka Consumer - KafkaConsumer subscription transition to unhealthy. consume loop"}
{"timestamp":"2023-01-30T21:04:10.6291230Z","level":"warning","message":"Kafka Consumer - Error consuming, retrying in 00:00:01: ...
{"timestamp":"2023-01-30T21:04:16.6295129Z","level":"info","message":"Kafka Consumer - KafkaConsumer subscription transition to healthy. after consume"}
```

> I guess if the new offset fields get marked as optional in DFS too, we may need to manually ignore messages that are missing the new fields.

That's a good point. We'd want to avoid sending a bogus broadcast. Right now, we're giving the new fields default values of zero instead of making them optional. To detect when they aren't set, I think it might be preferable to give them default values of -1, or something, just in case we want to special-case zero for "end of data". As a test, I tried to regenerate the avro schema C# class just now with an optional field, and it didn't make the field nullable, so I think we should prefer default values to optional fields in cases where we want to detect if something wasn't set.